src/distributions/weighted/mod.rs

```diff
@@ -125,9 +126,63 @@ impl<X: SampleUniform + PartialOrd> WeightedIndex<X> {
         if total_weight == zero {
             return Err(WeightedError::AllWeightsZero);
         }
-        let distr = X::Sampler::new(zero, total_weight);
+        let distr = X::Sampler::new(zero, total_weight.clone());
 
-        Ok(WeightedIndex { cumulative_weights: weights, weight_distribution: distr })
+        Ok(WeightedIndex { cumulative_weights: weights, total_weight, weight_distribution: distr })
+    }
+
+    /// Update a subset of weights, without changing the number of weights.
+    ///
+    /// Using this method instead of `new` might be more efficient if only a small number of
+    /// weights is modified. For weights that are `Copy`, no allocations are performed.
+    ///
+    /// In case of error, `self` is not modified.
+    pub fn update_weights(&mut self, new_weights: &[(usize, &X)]) -> Result<(), WeightedError>
+        where X: for<'a> ::core::ops::AddAssign<&'a X> +
+                 for<'a> ::core::ops::SubAssign<&'a X> +
+                 Clone +
+                 Default {
+        let zero = <X as Default>::default();
+
+        let mut total_weight = self.total_weight.clone();
+
+        for &(i, w) in new_weights {
+            if *w < zero {
+                return Err(WeightedError::InvalidWeight);
+            }
+            if i >= self.cumulative_weights.len() {
+                return Err(WeightedError::TooMany);
+            }
+
+            // Unfortunately, we will have to calculate the non-cumulative weight a second time, to
+            // avoid producing an invalid state of `self`.
+            let mut old_w = self.cumulative_weights[i].clone();
+            if i > 0 {
+                old_w -= &self.cumulative_weights[i - 1];
+            }
+
+            total_weight -= &old_w;
+            total_weight += w;
+        }
+        if total_weight == zero {
+            return Err(WeightedError::AllWeightsZero);
+        }
+
+        for &(i, w) in new_weights {
+            let mut old_w = self.cumulative_weights[i].clone();
+            if i > 0 {
+                old_w -= &self.cumulative_weights[i - 1];
+            }
+
+            for j in i..self.cumulative_weights.len() {
```

Comment:
This is `O(n*m)` where `n = cumulative_weights.len() - min_index; m = new_weights.len()`.

Instead we should sort the `new_weights` by index, then apply in-turn (like in `new`); this is `O(m*log(m) + n)`.

Also, we can just take `total_weight = cumulative_weights.last().unwrap()`.

Reply:
> Instead we should sort the new_weights by index, then apply in-turn (like in new); this is O(m*log(m) + n).

I'll look into this.

> Also, we can just take total_weight = cumulative_weights.last().unwrap().

I don't think so, the last cumulative weight is not stored in the vector. Or are you saying we should change it such that it is?

Reply:
Aha, `binary_search_by` is happy to return an index one-past-the-last-item, therefore the final weight is not needed. (And we have motive for not including the final weight: it guarantees we will never exceed the last index of the input `weights` list.)

Then yes, we need to store either the last weight or the total as an extra field.

Reply:
> Instead we should sort the new_weights by index, then apply in-turn (like in new); this is O(m*log(m) + n).

I implemented that. It's a bit messy, because the the index type might be unsigned.